### ISSUE TITLE
Training process and the derivation of MS-SSIM

### ISSUE THREAD

**Issue by duancaohui** [imodpasteur/ANNA-PALM#13]
Hi, I read your ANNA-PALM code , in particular for creating network and training. I have 2 question among them.
Loss function: super-res reconstruction error, the L1 norm and MS-SSIM, are implemented in `networks_tensorflow` just as a built-in function. However the implementation include only the forward  process, without backward. Tensorflow can't take the derivative of MS-SSIM with respect to output directly, TV regularization with the same question. How do you deal with this problem?

Training process: There are 3 networks in your ANNA-PALM:G, D, Q, 
![qq 20180718105955](https://user-images.githubusercontent.com/11900034/42856879-c1502e90-8a79-11e8-8a45-d47e59543706.png)

```
_losses = [gen_loss, gen_loss_L2, gen_loss_SSIM]
update_losses = ema.apply(list(set(_losses)))
incr_global_step = tf.assign(global_step, global_step+1)
train=tf.group(update_losses, incr_global_step, gen_train)
# Train
 fetches = {
                "train": self.model.train,
            }
results = self.sess.run(fetches, feed_dict={'dropout_prob:0': self.get_dropout_prob()}, options=options, run_metadata=run_metadata)
```
In the training process, there are no alternating update among the 3 networks?

**Comment by oeway** [comment]
Hi, thanks for your interest in understanding our method.
For your first question about the backward path of MS-SSIM and TV, since we used Tensorflow ops to implement the forward path, the backward path is done by Tensorflow itself. Almost deep learning library provide that feature, you don't need to care about the backward path if you only use ops for the composition.

Second, the alternation of the adversarial loss is designed in the network itself, [checkout here](https://github.com/imodpasteur/ANNA-PALM/blob/master/AnetLib/models/networks_tensorflow.py#L851). For the Low Resolution Estimator, there is no alternation, we train it simultaneously with other networks.

**Comment by duancaohui** [comment]
Thanks for your reply! 
The original implementation of L1 and MS-SSIM using caffe, in https://github.com/NVlabs/PL4NN, include forward and backward process. So I thought the MS-SSIM backwad is complicated for Tensorflow. In matconvnet, another deep learning library, it is also need to define forward and backward, the derivation of loss may be complex to handle, not like conv.

**Comment by oeway** [comment]
@duancaohui As mentioned, for all these deep learning library, including those you mentioned, if you use **only** their native ops to compose your op, layer or objective function, the backward path will be automatically computed.

**Comment by duancaohui** [comment]
I got it, thanks for your kindly reply.
This issue closed
